### PR TITLE
Add LineStringSegmentizeHaversine

### PR DESCRIPTION
**Pull request by JosiahParry**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/main/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

Supercedes https://github.com/georust/geo/pull/1106. 

This PR: 

- adds a new trait `LineStringSegmentizeHaversine` to segment a linestring using haversine distance instead of euclidean distance
  - this gives downstream users an option to segment linestrings on geographic as well as projected coordinates
- fixes a bug in `LineStringSegmentize` where `n - 1` linestrings were created for _very_ small distances
  - this was caused by precision in the 6-ish decimal place. It was fixed by subtracting `f64::EPSILON` from the calculated segment length that is passed to `densify()`
- turns the segmentize implementation into a macro.
  - the benefit of this is that segmentize for haversine does not need to duplicate code and other future implementations (hopefully geodesic) will not need more than a Densify and Length trait of the same type.

### CODE REVIEW

**Comment by JosiahParry** [comment]
Please let me know if this needs any additional work. :)

**Review by frewsxcv — APPROVED** [review]
lgtm! will see if others have thoughts before merging

**Comment by JosiahParry** [comment]
thanks for your help getting this across the line! Thank you!

**Pull request merged**
